packages/flutter/lib/src/widgets/editable_text.dart

```diff
@@ -5076,21 +5076,69 @@ class _ScribblePlaceholder extends WidgetSpan {
   }
 }
 
-/// A text boundary that uses code units as logical boundaries.
+/// A text boundary that uses code points as logical boundaries.
 ///
-/// This text boundary treats every character in input string as an utf-16 code
-/// unit. This can be useful when handling text without any grapheme cluster,
-/// e.g. password input in [EditableText]. If you are handling text that may
-/// include grapheme clusters, consider using [CharacterBoundary].
-class _CodeUnitBoundary extends TextBoundary {
-  const _CodeUnitBoundary(this._text);
+/// A code point represents a single character. This may be smaller than what is
+/// represented by a user-perceived character, or grapheme. For example, a
+/// single grapheme (in this case a Unicode extended grapheme cluster) like
+/// "👨‍👩‍👦" consists of five code points: the man emoji, a zero
+/// width joiner, the woman emoji, another zero width joiner, and the boy emoji.
+/// The [String] has a length of eight because each emoji consists of two code
+/// units.
+///
+/// Code units are the units by which Dart's String class is measured, which is
+/// encoded in UTF-16.
+///
+/// See also:
+///
+///  * [String.runes], which deals with code points like this class.
+///  * [String.characters], which deals with graphemes.
+///  * [CharacterBoundary], which is a [TextBoundary] like this class, but whose
+///    boundaries are graphemes instead of code points.
+class _CodePointBoundary extends TextBoundary {
+  const _CodePointBoundary(this._text);
 
   final String _text;
 
+  // Returns true if the given position falls in the center of a surrogate pair.
+  bool _breaksSurrogatePair(int position) {
+    assert(position > 0 && position < _text.length && _text.length > 1);
+    return _text.codeUnitAt(position) & 0xFC00 == 0xDC00
+        && _text.codeUnitAt(position + 1) & 0xFC00 == 0xD800;
+  }
+
   @override
-  int getLeadingTextBoundaryAt(int position) => position.clamp(0, _text.length); // ignore_clamp_double_lint
+  int? getLeadingTextBoundaryAt(int position) {
+    final int validatedPosition = math.min(math.max(position, 0), _text.length);
```

Comment:
for `position < 0` it needs to return `null` since there won't be any index that is less than or equal to `position` and points to a code point start position.

Reply:
Maybe something like this:
```dart
    return switch (position) {
      < 0                              => null,
      _ when position >= _text.length  => getLeadingTextBoundaryAt(_text.length - 1),
      _ when _isLowSurrogate(position) => position - 1,
      _                                => position,
    };
```
iirc it's guaranteed that `_text` isn't a malformed utf16 string

Reply:
I added the `null` cases and cleaned it up a bit.  I really wanted to include the cool new pattern matching but ended up going with `if`s haha.

I just tried using `tester.enterText` to enter `👨‍|👩‍�` and it called getLeadingTextBoundaryAt with the malformed string, so I left the handling in for mismatched surrogates.